File: Task/Determine-if-a-string-is-collapsible/F-Sharp/determine-if-a-string-is-collapsible.fs
// Collapse a String. Nigel Galloway: June 9th., 2020
//As per the task description a function which 'determines if a character string is collapsible' by testing if any consecutive characters are the same.
let isCollapsible n=n|>Seq.pairwise|>Seq.tryFind(fun(n,g)->n=g)
//As per the task description a function which 'if the string is collapsable, collapses the string (by removing immediately repeated characters).
let collapse n=match isCollapsible n with
                Some _->let i=Seq.head n
                        let fN=let mutable g=i in (fun n->if n=g then false else g<-n; true)
                        let g=System.String([|yield i;yield! Seq.tail n|>Seq.filter fN|])
                        printfn "<<<%s>>> (length %d) colapses to <<<%s>>> (length %d)" n n.Length g g.Length
               |     _->printfn "<<<%s>>> (length %d) does not colapse" n n.Length

collapse ""
collapse "\"If I were two-faced, would I be wearing this one?\" --- Abraham Lincoln "
collapse "..1111111111111111111111111111111111111111111111111111111111111117777888"
collapse "I never give 'em hell, I just tell the truth, and they think it's hell. "
collapse "                                                    --- Harry S Truman  "
collapse "withoutConsecutivelyRepeatedCharacters"

File: Task/Determine-if-a-string-is-collapsible/Delphi/determine-if-a-string-is-collapsible.pas
program Determine_if_a_string_is_collapsible;

{$APPTYPE CONSOLE}

uses
  System.SysUtils;

procedure collapsible(s: string);
var
  c, last: char;
  len: Integer;
begin
  writeln('old: <<<', s, '>>>, length = ', s.length);
  write('new: <<<');
  last := #0;
  len := 0;
  for c in s do
  begin
    if c <> last then
    begin
      write(c);
      inc(len);
    end;
    last := c;
  end;
  writeln('>>>, length = ', len, #10);
end;

begin
  collapsible('');
  collapsible('"If I were two-faced, would I be wearing this one?" --- Abraham Lincoln ');
  collapsible('..1111111111111111111111111111111111111111111111111111111111111117777888');
  collapsible('I never give ''em hell, I just tell the truth, and they think it''s hell. ');
  collapsible('                                                    --- Harry S Truman  ');
  readln;
end.
